File: quick_sort/quick_sort.py
import random


def sort(data):
    quick_sort(data, 0, len(data) - 1)


def quick_sort(data, idx_of_start, idx_end):
    if idx_of_start < idx_end:
        idx_of_partition = partition(data, idx_of_start, idx_end)

        quick_sort(data, idx_of_start, idx_of_partition - 1)
        quick_sort(data, idx_of_partition + 1, idx_end)


def partition(data, idx_of_start, idx_of_end):
    pivot = data[idx_of_end]

    j = idx_of_start
    for i in range(idx_of_start, idx_of_end):
        if data[i] < pivot:
            data[i], data[j] = data[j], data[i]

            j += 1

    data[j], data[idx_of_end] = data[idx_of_end], data[j]

    print(data)

    return j


if __name__ == '__main__':
    data = [random.randint(1, 20) for _ in range(20)]
    
    print("--- before ---")
    print(data)

    print("--- sort ---")
    sort(data)

    print("--- after ---")
    print(data)

File: quick_sort/quick_sort.go
package main

import (
	"fmt"
	"math/rand"
)

func createData() []int {
	data := make([]int, 20)

	for i := 0; i < 20; i++ {
		data[i] = rand.Intn(20)
	}

	return data
}

func sort(data []int) {
	quickSort(data, 0, len(data)-1)
}

func quickSort(data []int, idxOfStart, idxOfEnd int) {
	if idxOfStart < idxOfEnd {
		idxOfPartition := partition(data, idxOfStart, idxOfEnd)

		quickSort(data, idxOfStart, idxOfPartition-1)
		quickSort(data, idxOfPartition+1, idxOfEnd)
	}
}

func partition(data []int, idxOfStart, idxOfEnd int) int {
	pivot := data[idxOfEnd]

	i := idxOfStart
	for j := idxOfStart; j < idxOfEnd; j++ {
		if data[j] < pivot {
			data[i], data[j] = data[j], data[i]
			i += 1
		}
	}

	data[i], data[idxOfEnd] = data[idxOfEnd], data[i]

	fmt.Println(data)

	return i
}

func main() {
	data := createData()

	fmt.Println("--- before ---")
	fmt.Println(data)

	fmt.Println("--- sort ---")
	sort(data)

	fmt.Println("--- after ---")
	fmt.Println(data)
}
